src/nccl_ofi_net.c

```diff
@@ -393,14 +393,16 @@ static void get_hints(struct fi_info *hints, int req_gdr)
 		hints->domain_attr->mr_key_size = (size_t) ofi_nccl_mr_key_size();
 	}
 	else {
-		hints->caps = FI_TAGGED | FI_MSG | FI_REMOTE_COMM;
+		hints->caps = FI_TAGGED | FI_MSG;
 		/*
 		 * Set MR mode bits to indicate that application allows
 		 * registration of both local memory buffers
 		 */
 		hints->domain_attr->mr_mode = FI_MR_LOCAL;
 	}
 
+	hints->caps |= FI_LOCAL_COMM | FI_REMOTE_COMM;
```

Comment:
That makes sense.

There is something I didn't understand when reading the documentation of `FI_LOCAL_COMM`. The docs seem to imply that if this option is not set, the endpoint cannot communicate with other endpoints on the same node. If so, then it is not valid to not set `FI_LOCAL_COMM`, since we do need within-node communication. However, if not setting `FI_LOCAL_COMM` simply means that within-node communications will go through the EFA device and not through SHM, then we probably don't need to set this flag. I don't think within-node communication is performance-critical for us, since it just needs to be faster than inter-node.

@a-szegel @shijin-aws do you have any thoughts on this? I agree that if setting `FI_LOCAL_COMM` makes local reads go through SHM instead of the EFA device that would be undesirable.

Reply:
I think even with this PR, shm won't be used for tx for nccl plugin, as nccl plugin will set `efa_ep->cuda_api_permitted`  to be false, and shm cannot be used for this setting

Reply:
All of this is noise.  Eric is right; for the local flush or any local operations, we need to request `FI_LOCAL_COMM`.  Getting the actual data in and out of the GPU using EFA is orthogonal to that and should not be coupled into the `FI_LOCAL_COMM` flag.  It's possible other things in the EFA provider are broken, but this is the right usage of the Libfabric API.